src/components/KeyboardAvoidingView/index.android.tsx

```diff
@@ -0,0 +1,135 @@
+import React, {forwardRef, useCallback, useMemo, useState} from 'react';
+import type {LayoutRectangle, View, ViewProps} from 'react-native';
+import {useKeyboardContext, useKeyboardHandler} from 'react-native-keyboard-controller';
+import Reanimated, {interpolate, runOnUI, useAnimatedStyle, useDerivedValue, useSharedValue} from 'react-native-reanimated';
+import {useSafeAreaFrame} from 'react-native-safe-area-context';
+import type {KeyboardAvoidingViewProps} from './types';
+
+const useKeyboardAnimation = () => {
+    const {reanimated} = useKeyboardContext();
+
+    // calculate it only once on mount, to avoid `SharedValue` reads during a render
+    const [initialHeight] = useState(() => -reanimated.height.value);
+    const [initialProgress] = useState(() => reanimated.progress.value);
+
+    const heightWhenOpened = useSharedValue(initialHeight);
+    const height = useSharedValue(initialHeight);
+    const progress = useSharedValue(initialProgress);
+    const isClosed = useSharedValue(initialProgress === 0);
+
+    useKeyboardHandler(
+        {
+            onStart: (e) => {
+                'worklet';
+
+                progress.value = e.progress;
+                height.value = e.height;
+
+                if (e.height > 0) {
+                    // eslint-disable-next-line react-compiler/react-compiler
+                    isClosed.value = false;
+                    heightWhenOpened.value = e.height;
+                }
+            },
+            onEnd: (e) => {
+                'worklet';
+
+                isClosed.value = e.height === 0;
+
+                height.value = e.height;
+                progress.value = e.progress;
+            },
+        },
+        [],
+    );
+
+    return {height, progress, heightWhenOpened, isClosed};
+};
+
+const defaultLayout: LayoutRectangle = {
+    x: 0,
+    y: 0,
+    width: 0,
+    height: 0,
+};
+
+/**
+ * View that moves out of the way when the keyboard appears by automatically
+ * adjusting its height, position, or bottom padding.
+ *
+ * This `KeyboardAvoidingView` acts as a backward compatible layer for the previous Android behavior (prior to edge-to-edge mode).
+ * We can use `KeyboardAvoidingView` directly from the `react-native-keyboard-controller` package, but in this case animations are stuttering and it's better to handle as a separate task.
```

Comment:
Unrelated, but is this something that might be resolved in the future?

Reply:
> This `KeyboardAvoidingView` acts as a backward compatible layer for the previous Android behavior (prior to edge-to-edge mode).

Reply:
> Unrelated, but is this something that might be resolved in the future?

Yes, definitely. The problem is that if we start to use `KeyboardAvoidingView` directly from `react-native-keyboard-controller`, then on some screens animations performance will be very poor.

I tried to quickly resolve that, but I realized, that it will take more time, than I initially expected, that's why I added this compat layer. Once this PR merged I'll be happy to track a new issue and start to work on removing this compat layer and replace it with original `KeyboardAvoidingView` implementation (we just need to figure out what causes such severe performance drops, but again better to handle it separately).